src/tools/ceph_objectstore_tool.cc

```diff
@@ -1869,45 +1685,73 @@ int ObjectStoreTool::do_import(ObjectStore *store, OSDSuperblock& sb,
     switch(type) {
     case TYPE_OBJECT_BEGIN:
       ceph_assert(found_metadata);
-      ret = get_object(store, driver, mapper, coll, ebl, curmap, &skipped_objects);
+      ret = get_object(store, driver, mapper, coll, ebl, ms.osdmap,
+		       &skipped_objects);
       if (ret) return ret;
       break;
     case TYPE_PG_METADATA:
-      ret = get_pg_metadata(store, ebl, ms, sb, curmap, pgid);
+      ret = get_pg_metadata(store, ebl, ms, sb, pgid);
       if (ret) return ret;
       found_metadata = true;
 
-      // make sure there are no conflicting merges
-      {
+      if (pgid != ms.info.pgid) {
+	cerr << "specified pgid " << pgid << " does not match import file pgid "
+	     << ms.info.pgid << std::endl;
+	return -EINVAL;
+      }
+
+      // make sure there are no conflicting splits or merges
+      if (ms.osdmap.have_pg_pool(pgid.pgid.pool())) {
 	auto p = pg_num_history.pg_nums.find(pgid.pgid.m_pool);
-	if (p != pg_num_history.pg_nums.end()) {
-	  unsigned pg_num = ms.osdmap.get_pg_num(pgid.pgid.m_pool);
+	if (p != pg_num_history.pg_nums.end() &&
+	    !p->second.empty()) {
+	  unsigned pg_num = ms.osdmap.get_pg_num(pgid.pgid.pool());
 	  for (auto q = p->second.lower_bound(ms.map_epoch);
 	       q != p->second.end();
 	       ++q) {
-	    pg_t parent;
-	    if (pgid.pgid.is_merge_source(pg_num, q->second, &parent)) {
-	      cerr << "PG " << pgid.pgid << " merge source in epoch "
-		   << q->first << " pg_num " << pg_num
-		   << " -> " << q->second << std::endl;
-	      return 12;
+	    unsigned new_pg_num = q->second;
+	    cout << "pool " << pgid.pgid.pool() << " pg_num " << pg_num
+		 << " -> " << new_pg_num << std::endl;
+
+	    // check for merge target
+	    spg_t target;
+	    if (pgid.is_merge_source(pg_num, new_pg_num, &target)) {
+	      // FIXME: this checks assumes the OSD's PG is at the OSD's
+	      // map epoch; it could be, say, at *our* epoch, pre-merge.
+	      coll_t coll(target);
+	      if (store->collection_exists(coll)) {
+		cerr << "pgid " << pgid << " merges to target " << target
+		     << " which already exists" << std::endl;
+		return 12;
+	      }
 	    }
-	    if (pgid.pgid.is_merge_target(pg_num, q->second)) {
-	      cerr << "PG " << pgid.pgid << " merge target in epoch "
-		   << q->first << " pg_num " << pg_num
-		   << " -> " << q->second << std::endl;
-	      return 12;
+
+	    // check for split children
+	    set<spg_t> children;
+	    if (pgid.is_split(pg_num, new_pg_num, &children)) {
+	      for (auto child : children) {
+		coll_t coll(child);
+		if (store->collection_exists(coll)) {
+		  cerr << "pgid " << pgid << " splits to " << children
+		       << " and " << child << " exists" << std::endl;
+		  return 12;
+		}
+	      }
 	    }
-	    pg_num = q->second;
+	    pg_num = new_pg_num;
 	  }
 	}
+      } else {
+	cout << "pool " << pgid.pgid.pool() << " doesn't existing, not checking"
```

Comment:
s/existing/exist/

Reply:
@liewegas As stated above why doesn't this say "Pool ## no longer exists" and fail like the old code?